src/lib/crypto/snarky_tests/snarky_tests.ml

```diff
@@ -434,6 +434,138 @@ let api_tests =
   ; ("compile monadic API", `Quick, MonadicAPI.get_hash_of_circuit)
   ]
 
+module Improper_calls = struct
+  let input_typ = Impl.Typ.tuple2 Impl.Field.typ Impl.Field.typ
+
+  let return_typ = Impl.Typ.unit
+
+  let use_circuit_functions a b : unit =
+    let ab = Impl.Field.(a + b) in
+    Impl.Field.Assert.non_zero ab ;
+    ()
+
+  let use_prover_functions a b : unit =
+    let a = Impl.As_prover.read_var a in
+    let b = Impl.As_prover.read_var b in
+    let ab = Impl.Field.Constant.(a + b) in
+    assert (not (Impl.Field.Constant.(compare ab one) = 0)) ;
+    ()
+
+  let random_input = Impl.Field.Constant.(random (), random ())
+
+  let run_circuit_as_prover circuit : unit =
+    let f (inputs : Impl.Proof_inputs.t) _ = inputs in
+    let compiled =
+      Impl.generate_witness_conv ~f ~input_typ ~return_typ circuit
+    in
+    let input = random_input in
+    let _b = compiled input in
+    ()
+
+  let run_circuit_as_circuit circuit : unit =
+    let cs : Impl.R1CS_constraint_system.t =
+      Impl.constraint_system ~input_typ ~return_typ circuit
+    in
+    let _digest = Md5.to_hex (Impl.R1CS_constraint_system.digest cs) in
+    ()
+
+  module Tests = struct
+    let circuit_function_inside_circuit_inside_prover () =
+      let inner_circuit ((a, b) : Impl.Field.t * Impl.Field.t) () : unit =
+        use_circuit_functions a b ; ()
+      in
+      let circuit ((_a, _b) : Impl.Field.t * Impl.Field.t) () =
+        Impl.as_prover (fun _ ->
+            run_circuit_as_circuit inner_circuit ;
+            () )
+      in
+      run_circuit_as_prover circuit ;
```

Comment:
I this test the inner circuit is used to build a constraint system instead of witness generation, first two are still relevant but I will implement them in the other tests that makes witness generation.
For third point given that it is a statement I can just wrap in an if and compare hashes.